core/commands/dag/dag.go

```diff
@@ -241,3 +261,248 @@ var DagResolveCmd = &cmds.Command{
+				} else if err := node.Pinning.Flush(req.Context); err != nil {
+					ret.PinErrorMsg = err.Error()
+				}
+
+				if ret.PinErrorMsg != "" {
+					failedPins++
+				}
+
+				if err := res.Emit(&CarImportOutput{Root: ret}); err != nil {
+					return err
+				}
+			}
+
+			if failedPins > 0 {
+				return fmt.Errorf(
+					"unable to pin all roots: %d out of %d failed",
+					failedPins,
+					len(roots),
+				)
+			}
+		}
+
+		return nil
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, event *CarImportOutput) error {
+
+			silent, _ := req.Options[silentOptionName].(bool)
+			if silent {
+				return nil
+			}
+
+			enc, err := cmdenv.GetLowLevelCidEncoder(req)
+			if err != nil {
+				return err
+			}
+
+			if event.Root.PinErrorMsg != "" {
+				event.Root.PinErrorMsg = fmt.Sprintf("FAILED: %s", event.Root.PinErrorMsg)
+			} else {
+				event.Root.PinErrorMsg = "success"
+			}
+
+			if !event.Root.PresentInCar {
+				event.Root.PinErrorMsg += " (root specified in .car header without its data)"
+			}
+
+			_, err = fmt.Fprintf(
+				w,
+				"Pinned root\t%s\t%s\n",
+				enc.Encode(event.Root.Cid),
+				event.Root.PinErrorMsg,
+			)
+			return err
+		}),
+	},
+}
+
+func importWorker(req *cmds.Request, re cmds.ResponseEmitter, api iface.CoreAPI, ret chan importResult) {
+
+	// this is *not* a transaction
+	// it is simply a way to relieve pressure on the blockstore
+	// similar to pinner.Pin/pinner.Flush
+	batch := ipld.NewBatch(req.Context, api.Dag())
+
+	roots := make(map[cid.Cid]bool)
+
+	it := req.Files.Entries()
+	for it.Next() {
+
+		file := files.FileFromEntry(it)
+		if file == nil {
+			ret <- importResult{err: errors.New("expected a file handle")}
+			return
+		}
+
+		car, err := gocar.NewCarReader(file)
+		if err != nil {
+			ret <- importResult{err: err}
+			return
+		}
+
+		// Be explicit here, until the spec is finished
+		if car.Header.Version != 1 {
+			ret <- importResult{err: errors.New("only car files version 1 supported at present")}
+			return
+		}
+
+		for _, c := range car.Header.Roots {
+			if _, exists := roots[c]; !exists {
+				roots[c] = false
+			}
+		}
+
+		for {
+			block, err := car.Next()
+			if err != nil && err != io.EOF {
+				ret <- importResult{err: err}
+				return
+			} else if block == nil {
+				break
+			}
+
+			// the double-decode is suboptimal, but we need it for batching
+			nd, err := ipld.Decode(block)
+			if err != nil {
+				ret <- importResult{err: err}
+				return
+			}
+
+			if err := batch.Add(req.Context, nd); err != nil {
+				ret <- importResult{err: err}
+				return
+			}
+
+			// encountered something known to be a root, for the first time
+			if seen, exists := roots[nd.Cid()]; exists && !seen {
+				roots[nd.Cid()] = true
+			}
+		}
+
+		// every single file in it() is already open before we start
+		// just close here sooner rather than later for neatness
+		// this won't/can't help with not running out of handles
+		file.Close()
```

Comment:
should these be closed in error cases as well? is there a deferred close scheduled as well?

Reply:
@Stebalien thoughts?

Reply:
It may not matter in practice, but we probably _should_ anyways.

Reply:
Rewired. It's not pretty in diff-form ( hideous actually ), but the actual resulting code is ok.